### DISCUSSION TITLE
Does the redis scaler support redis cluster?

### DISCUSSION THREAD

**Discussion by mboutet** [kedacore/keda#872] [General]
https://redis.io/topics/cluster-spec

Thanks

**Comment by tomkerkhove** [comment]
Today we only support scaling on [Redis List](https://keda.sh/docs/1.4/scalers/redis-lists/) and [Redis Streams](https://github.com/kedacore/keda/pull/765) is on the way, but not based on redis cluster info.

What's your scenario exactly? Based on that we can see if it's a good addition to the suite (which I think it is)

**Reply by mboutet** [reply]
Redis cluster is a type of redis deployment using a multi-master topology where keys are distributed according to hash slots. The client needs to be aware of this topology in order to interact with the cluster since a redis node can reply with a `ASK` or `MOVE` to indicate to the client which redis node to read from or write to (commands are not proxies by the redis nodes, it is the client responsibility to query the right nodes).

In our case, we use redis cluster in our stack to provide high availability and greater scalability.

When using redis cluster, we typically provide a list of startup redis nodes to the client such that it can query each node and cache the cluster topology in memory. But I think that you can also provide the host of a single node and start from there. Nevertheless, the client still needs to know what to do with `ASK` or `MOVE` directives.

**Reply by tomkerkhove** [reply]
Thanks for the info, so basically you are more asking if our current scalers are capable of working with such a topology if I get it correctly?

I do not know that, maybe @abhirockzz knows?

**Reply by mboutet** [reply]
Yes, you got it. It looks like the [`go-redis`](https://pkg.go.dev/github.com/go-redis/redis/v8?tab=doc#NewClusterClient) client support the cluster topology out of the box. I also know that redis cluster is offered on [Azure](https://docs.microsoft.com/en-us/azure/azure-cache-for-redis/cache-how-to-premium-clustering) and [AWS](https://aws.amazon.com/blogs/database/work-with-cluster-mode-on-amazon-elasticache-for-redis/). In my case, I am deploying redis cluster directly in k8s in statefulsets with azure disks backed PVs for persistence.

**Comment by abhirockzz** [comment]
Thanks for the details @mboutet ! Right now, the Redis (List) scaler does not support a cluster topology - it's just a normal client connection for now using `redis.NewClient`

**Comment by abhirockzz** [comment]
That being said, it should be possible to create a clustered client based on a flag specified in the `ScaledObject` trigger metadata

**Reply by zroubalik** [reply]
@abhirockzz do you feel that's something that is worth adding to current [Readis Streams PR](https://github.com/kedacore/keda/pull/765) (I was just about to merge this one)?
Or is it something that could be unified on Redis Scalers?

**Reply by abhirockzz** [reply]
This applies to Redis in general, so it makes sense to keep it in a common place @zroubalik

**Reply by zroubalik** [reply]
@abhirockzz ok, thanks!

@mboutet Please create an issue if you feel this is something that you would like to see in KEDA

**Reply by mboutet** [reply]
I created a feature request: https://github.com/kedacore/keda/issues/883.

Thank you
